include/enotify.php

```diff
@@ -587,6 +603,102 @@ function notification($params)
 	return false;
 }
 
+function notification_from_array(array $notification)
+{
+	if ($notification['type'] == Post\UserNotification::NOTIF_NONE) {
+		return false;
+	}
+
+	$params = [];
+	$params['verb'] = Verb::getByID($notification['vid']);
+
+	$params['uid']   = $notification['uid'];
+	$params['otype'] = Notification\ObjectType::ITEM;
+
+	$user = User::getById($notification['uid']);
+
+	$params['notify_flags'] = $user['notify-flags'];
+	$params['language']     = $user['language'];
+	$params['to_name']      = $user['username'];
+	$params['to_email']     = $user['email'];
+
+	// from here on everything is in the recipients language
+	$l10n = DI::l10n()->withLang($user['language']);
+
+	$contact = Contact::getById($notification['actor-id'], ['url', 'name', 'photo']);
+	if (DBA::isResult($contact)) {
+		$params['source_link']  = $params['origin_link']  = $contact['url'];
+		$params['source_name']  = $params['origin_name']  = $contact['name'];
+		$params['source_photo'] = $params['origin_photo'] = $contact['photo'];
+	}
+
+	$item = Post::selectFirstForUser($notification['uid'], Item::ITEM_FIELDLIST,
+		['uid' => [0, $notification['uid']], 'uri-id' => $notification['target-uri-id'], 'deleted' => false],
+		['order' => ['uid' => true]]);
+	if (empty($item)) {
+		return false;
+	}
+
+	$params['item']   = $item;
+	$params['parent'] = $item['parent'];
+	$params['link']   = DI::baseUrl() . '/display/' . urlencode($item['guid']);
+
+	$subjectPrefix = $l10n->t('[Friendica:Notify]');
+
+	if (Post\ThreadUser::getIgnored($notification['parent-uri-id'], $notification['uid'])) {
+		Logger::info('Thread is ignored', ['parent-uri-id' => $notification['parent-uri-id']]);
+		return false;
+	}
+
+	// Check to see if there was already a tag notify or comment notify for this post.
+	// If so don't create a second notification
+	$condition = ['type' => [Notification\Type::TAG_SELF, Notification\Type::COMMENT, Notification\Type::SHARE],
+		'link' => $params['link'], 'uid' => $notification['uid']];
+	if (DBA::exists('notify', $condition)) {
+		return false;
+	}
+
+	$content = Plaintext::getPost($item, 70);
+	if (!empty($content['text'])) {
+		$title = '"' . trim(str_replace("\n", " ", $content['text'])) . '"';
+	} else {
+		$title = $item['title'];
+	}
+
+	// Some mail software relies on subject field for threading.
+	// So, we cannot have different subjects for notifications of the same thread.
+	// Before this we have the name of the replier on the subject rendering
+	// different subjects for messages on the same thread.
```

Comment:
I don’t understand this comment block, it says that we can’t have different notification email subjects for the same thread before a conditional block where the subject depends on the notification type and the remote contact?

Reply:
This is just some c&p. I haven't made my thoughts about this.This whole file should be replaced sooner than later - and then we have to think about the functionality.